### DISCUSSION TITLE
JOIN subquery IN with a limit inside with MySql  Strange behaviour

### DISCUSSION THREAD

**Discussion by Claod44** [symfony/symfony#49595] [Q&A]
### Symfony version(s) affected

6.1.12

### Description

I want to paginate Articles with their comments.
I'm supposed to get 2 articles with their comments.
With mysql I can't 

```sql
SELECT * FROM tableOrSelect WHERE tableOrSelect .column IN (select .... LIMIT 2 ...)
```

This version of MySQL doesn't yet support 'LIMIT & IN/ALL/ANY/SOME subquery

So the workaround is to use an inner join with the subselect ... really hard to make with the queryBuilder 
Those two ones are returning almost the same datas :

This :

```sql
SELECT * FROM (SELECT * FROM article as filtredArticles WHERE filtredArticles.role_requis IN ("PUBLIC_ACCESS","ROLE_USER","ROLE_ADMIN") ORDER BY filtredArticles.id ASC LIMIT 2 OFFSET 2) as article LEFT JOIN comment on article.id = comment.article_id;
```

OR
```sql
SELECT * FROM comment INNER JOIN (SELECT * FROM article as filtredArticles WHERE filtredArticles.role_requis IN ("PUBLIC_ACCESS","ROLE_USER","ROLE_ADMIN") ORDER BY filtredArticles.id ASC LIMIT 2 OFFSET 2) as filteredArticles on filteredArticles.id = comment.article_id;
```

When we convert this in DQL the limit and the offset are gone.... instead of being in the subrequest.

**I don't know if it's correct**

```php
$articleSelect = $this->createQueryBuilder('fa')->select('filteredArticle.id')->from(Article::class, 'filteredArticle');
        $articleSelect->andWhere($articleSelect->expr()->in('filteredArticle.role_requis', Roles::getGrantedRoles($user)))
            ->orderBy('filteredArticle.id', 'ASC')
            ->setMaxResults($limitNumberByPage)
            ->setFirstResult(($currentPageNumber * $limitNumberByPage) - $limitNumberByPage);

        $commentSelect = $this->createQueryBuilder('article')->join('article.comments', 'comments')->addSelect('comments');
        $commentSelect->where($commentSelect->expr('comments')->in('comments.article', $articleSelect->getDQL()));
```

### How to reproduce

Debug the string generated with a mysql database,
Structure  :

Article table with id, body, title... 
Comment table id, content, date...  linked to an article by a foreign key.

1 article may have 0..many comment(s)
1 comment must have 1 article

### Possible Solution

Easier way to join a subselect with a chained left join... I don't really know

### Additional Context

The limit and the offset are mandatory because of the paginator.

EDITED : Changed the SQL by a smaller version, removed wrong DQL conversion.

**Comment by Claod44** [comment]
Sorry to up) but ...

We cannot use Limit, We cannot use Offset, We cannot use ROW_NUMBER ...

How to paginate with 1 to many then please

**Comment by lyrixx** [comment]
Pagination with join is hard indeed! Did you try to use an OSS pagination like [KnpPaginatorBundle](https://github.com/KnpLabs/KnpPaginatorBundle) ?

If it does not work, but you manage to write a SQL query, I invite you to execute the raw SQL query, then hydrate everything with the [Result Set Mapper of doctrine](https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/reference/native-sql.html)
